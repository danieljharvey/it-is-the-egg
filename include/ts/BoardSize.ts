export class BoardSize {
  width: number;
  height: number;

  minSize: number = 5;
  maxSize: number = 40;

  constructor(size: number) {
    if (size < this.minSize) size = this.minSize;
    if (size > this.maxSize) size = this.maxSize;
    this.width = this.height = size;
  }

  grow(): void {
    if (this.width < this.maxSize) this.width++;
    if (this.height < this.maxSize) this.height++;
  }

  shrink(): void {
    if (this.width > this.minSize) this.width--;
    if (this.height > this.minSize) this.height--;
  }

  getData() {
    return {
      width: this.width,
      height: this.height
    };
  }
}
