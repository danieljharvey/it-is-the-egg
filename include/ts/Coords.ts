import { Record } from "immutable";
import { Utils } from "./Utils";

const SPRITE_SIZE: number = 64;

type CoordsParams = {
  x?: number,
  y?: number,
  offsetX?: number,
  offsetY?: number
};

export class Coords extends Record({ x: 0, y: 0, offsetX: 0, offsetY: 0 }) {
  public x: number;
  public y: number;
  public offsetX: number;
  public offsetY: number;

  constructor(params?: CoordsParams) {
      params ? super(params) : super();
  }

  public modify(values: CoordsParams) {
      return this.merge(values) as this;
  }

  public getActualPosition() {
    const fullX: number = this.x * SPRITE_SIZE + this.offsetX;
    const fullY: number = this.y * SPRITE_SIZE + this.offsetY;
    return {
      fullX,
      fullY
    };
  }
}
/*
export class Coords {
  public readonly x: number = 0;
  public readonly y: number = 0;
  public readonly offsetX: number = 0;
  public readonly offsetY: number = 0;

  constructor(x: number, y: number, offsetX: number = 0, offsetY: number = 0) {
    this.x = Math.floor(x) as number;
    this.y = Math.floor(y) as number;
    this.offsetX = offsetX as number;
    this.offsetY = offsetY as number;
  }

  public getActualPosition() {
    const fullX: number = this.x * SPRITE_SIZE + this.offsetX;
    const fullY: number = this.y * SPRITE_SIZE + this.offsetY;
    return {
      fullX,
      fullY
    };
  }

  public equals(otherCoords: Coords) {
    if (this.x !== otherCoords.x) {
      return false;
    }
    if (this.y !== otherCoords.y) {
      return false;
    }
    if (this.offsetX !== otherCoords.offsetX) {
      return false;
    }
    if (this.offsetY !== otherCoords.offsetY) {
      return false;
    }
    return true;
  }

  public modify(params: object): Coords {
    const newParams = (Object as any).assign({}, this, params);
    return new Coords(
      newParams.x,
      newParams.y,
      newParams.offsetX,
      newParams.offsetY
    );
  }
}
*/