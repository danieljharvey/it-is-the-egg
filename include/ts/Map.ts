import { Board } from "./Board";
import { BoardSize } from "./BoardSize";
import { Coords } from "./Coords";
import { Player } from "./Player";
import { Tile } from "./Tile";
import { TileSet } from "./TileSet";
import { Utils } from "./Utils";

// map is just a class full of functions that is created for manipulating the board
// does not contain any state of it's own

export class Map {
  public boardSize: BoardSize;

  protected tileSet: TileSet;

  protected board: Board;

  constructor(tileSet: TileSet, boardSize: BoardSize) {
    this.tileSet = tileSet;
    this.boardSize = boardSize;
  }

  public shrinkBoard(board: Board) : Board {
    this.boardSize = this.boardSize.shrink();
    return this.correctBoardSizeChange(board, this.boardSize);
  }

  public growBoard(board: Board) : Board {
    this.boardSize = this.boardSize.grow();
    return this.correctBoardSizeChange(board, this.boardSize);
  }

  // board is current board
  // boardSize is intended board size
  // returns new Board
  public correctBoardSizeChange(board: Board, boardSize: BoardSize) : Board {
    const newBoard = [];

    const currentWidth = board.getLength();

    const currentHeight = currentWidth;

    for (let x = 0; x < boardSize.width; x++) {
      newBoard[x] = [];
      for (let y = 0; y < boardSize.height; y++) {
        if (x < currentWidth && y < currentHeight) {
          // using current board
          const tile = board.getTile(x,y);
          newBoard[x][y] = tile;
        } else {
          // adding blank tiles
          const tile = this.cloneTile(1);
          newBoard[x][y] = tile;
        }
      }
    }
    return new Board(newBoard);
  }

  public generateBlankBoard() : Board {
    const board = [];

    for (let x = 0; x < this.boardSize.width; x++) {
      board[x] = [];
      for (let y = 0; y < this.boardSize.height; y++) {
        const blankTile = this.cloneTile(1);
        const positionedTile = blankTile.modify({
          x,
          y
        });
        board[x][y] = positionedTile;
      }
    }
    return new Board(board);
  }

  public correctForOverflow(coords: Coords): Coords {
    return Utils.correctForOverflow(coords, this.boardSize);
  }

  // is intended next tile empty / a wall?
  public checkTileIsEmpty(board: Board, x, y) : boolean {
    const tile = this.getTile(board, x, y);
    return tile.background;
  }

  public getTileWithCoords(board: Board, coords: Coords): Tile {
    const fixedCoords = this.correctForOverflow(coords);
    const { x, y } = fixedCoords;
    return board.getTile(x,y);
  }

  public changeTile(board: Board, coords: Coords, tile: Tile) : Board {
    return board.modify(coords.x, coords.y, tile);
  }

  public rotatePlayer(player: Player, clockwise) : Player {
    const newCoords = this.translateRotation(player.coords, clockwise);

    let direction = player.direction;
    // if player is still, nudge them in rotation direction
    if (direction === 0) {
      if (clockwise) {
        direction = 1;
      } else {
        direction = -1;
      }
    }

    return player.modify({
      coords: newCoords.modify({
        offsetX: 0,
        offsetY: 0
      }),
      direction
    });
  }

  public cloneTile(id): Tile {
    const prototypeTile = this.getPrototypeTile(id);
    return new Tile(prototypeTile); // create new Tile object with these
  }

  public getRandomTile(tiles) : Tile {
    const randomProperty = obj => {
      const randomKey = Utils.getRandomObjectKey(obj);
      return this.cloneTile(randomKey);
    };

    const theseTiles = this.tileSet.getTiles();
    (Object as any).entries(theseTiles).filter(([key, tile]) => {
      if (tile.dontAdd) {
        delete theseTiles[key];
      }
      return true;
    });
    return randomProperty(theseTiles);
  }

  // swap two types of tiles on map (used by pink/green switching door things)
  public switchTiles(board: Board, id1, id2) : Board {
    const tiles = board.getAllTiles();
    return tiles.reduce((currentBoard, tile) => {
      if (tile.id === id1) {
        const newTile = this.cloneTile(id2); 
        return currentBoard.modify(tile.x, tile.y, newTile);
      } else if (tile.id === id2) {
        const newTile = this.cloneTile(id1); 
        return currentBoard.modify(tile.x, tile.y, newTile);
      }
      return currentBoard;
    }, board);
  }

  // find random tile of type that is NOT at currentCoords
  public findTile(board: Board, currentCoords: Coords, id) : Tile | boolean {
    const tiles = board.getAllTiles();
    const teleporters = tiles.filter(tile => {
      if (tile.x === currentCoords.x && tile.y === currentCoords.y) {
        return false;
      }
      return tile.id === id;
    });
    if (teleporters.length === 0) {
      return false; // no options
    }
    const newTile = teleporters[Math.floor(Math.random() * teleporters.length)];
    return newTile;
  }

  // rotates board, returns new board and new renderAngle
  // really should be two functions
  public rotateBoard(board: Board, clockwise: boolean) {
    const tiles = board.getAllTiles();

    const width = board.getLength() - 1;
    const height = board.getLength() - 1;

    const rotatedBoard = tiles.reduce((currentBoard, tile) => {
      const coords = new Coords({ x: tile.x, y: tile.y });
      const newCoords = this.translateRotation(coords, clockwise);
      const newTile = tile.modify({
        x: newCoords.x,
        y: newCoords.y
      });
      return currentBoard.modify(newCoords.x, newCoords.y, newTile);
    }, board);

    return rotatedBoard;
  }

  public changeRenderAngle(renderAngle: number, clockwise: boolean) {
    let newRenderAngle;
    if (clockwise) {
      newRenderAngle = renderAngle + 90;
      if (newRenderAngle > 360) {
        newRenderAngle = newRenderAngle - 360;
      }
      return newRenderAngle
    }

    newRenderAngle = renderAngle - 90;
    if (newRenderAngle < 0) {
      newRenderAngle = 360 + newRenderAngle;
    }
    return newRenderAngle;
  }


  public makeBoardFromArray(boardArray: Tile[][] = []) : Board {
    const newBoard = boardArray.map((column, mapX) => {
      return column.map((item, mapY) => {
        const newTile = this.cloneTile(item.id);
        return newTile.modify({
          x: mapX,
          y: mapY
        });
      });
    });
    return new Board(newBoard);
  }

  protected getTile(board: Board, x: number, y: number) {
    const coords = new Coords({ x, y });
    return this.getTileWithCoords(board, coords);
  }

  protected generateRandomBoard(boardSize: BoardSize) : Board {
    const boardArray = [];

    for (let x = 0; x < boardSize.width; x++) {
      boardArray[x] = [];
      for (let y = 0; y < boardSize.height; y++) {
        const blankTile = this.getRandomTile(this.tileSet.getTiles());
        const positionedTile = blankTile.modify({
          x,
          y
        });
        boardArray[x][y] = blankTile;
      }
    }
    return new Board(boardArray);
  }

  protected getPrototypeTile(id) : object {
    return this.tileSet.getTile(id);
  }

  protected translateRotation(coords: Coords, clockwise: boolean): Coords {
    const width = this.boardSize.width - 1;
    const height = this.boardSize.height - 1;

    if (clockwise) {
      // 0,0 -> 9,0
      // 9,0 -> 9,9
      // 9,9 -> 0,9
      // 0,9 -> 0,0
      return coords.modify({
        x: width - coords.y,
        y: coords.x
      });
    } else {
      // 0,0 -> 0,9
      // 0,9 -> 9,9
      // 9,9 -> 9,0
      // 9,0 -> 0,0
      return coords.modify({
        x: coords.y,
        y: height - coords.x
      });
    }
  }
}
