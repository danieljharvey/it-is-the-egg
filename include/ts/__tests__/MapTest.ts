import { Coords } from "../Coords";
import { Map } from "../Map";
import { TileSet } from "../TileSet";
import { BoardSize } from "../BoardSize";

function configureTileSetMock() {
  const tileSet = jest.fn(); // create mock that does nothing
  tileSet.getTile = id => {
    return {
      id: id,
      title: "made up tile"
    };
  };
  return tileSet;
}

test("Translate rotation", () => {
  const rotateData = [
    { inX: 0, inY: 0, clockwise: true, outX: 9, outY: 0 },
    { inX: 9, inY: 0, clockwise: true, outX: 9, outY: 9 },
    { inX: 9, inY: 9, clockwise: true, outX: 0, outY: 9 },
    { inX: 0, inY: 9, clockwise: true, outX: 0, outY: 0 },
    { inX: 0, inY: 0, clockwise: false, outX: 0, outY: 9 },
    { inX: 0, inY: 9, clockwise: false, outX: 9, outY: 9 },
    { inX: 9, inY: 9, clockwise: false, outX: 9, outY: 0 },
    { inX: 9, inY: 0, clockwise: false, outX: 0, outY: 0 }
  ];

  const boardSize = new BoardSize(10);

  const map = new Map(undefined, boardSize);

  rotateData.map(data => {
    const expected = new Coords({ x: data.outX, y: data.outY });

    const coords = new Coords({ x: data.inX, y: data.inY });
    const result = map.translateRotation(coords, data.clockwise);
    return expect(result).toEqual(expected);
  });
});

test("Correct board size with shrinking", () => {
  const board = [
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0]
  ];

  const boardSize = {
    width: 4,
    height: 4
  };

  const expected = [[0, 1, 0, 0], [0, 1, 0, 0], [0, 1, 0, 0], [0, 1, 0, 0]];

  const map = new Map();

  const result = map.correctBoardSizeChange(board, boardSize);
  expect(result).toEqual(expected);
});

test("Correct board size with shrinking", () => {
  const board = [
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0]
  ];

  const boardSize = {
    width: 6,
    height: 6
  };

  const tileSet = new TileSet();
  const map = new Map(tileSet);

  const tile = map.cloneTile(1);

  const expected = [
    [0, 1, 0, 0, 0, tile],
    [0, 1, 0, 0, 0, tile],
    [0, 1, 0, 0, 0, tile],
    [0, 1, 0, 0, 0, tile],
    [0, 1, 0, 0, 0, tile],
    [tile, tile, tile, tile, tile, tile]
  ];

  const result = map.correctBoardSizeChange(board, boardSize);
  expect(result).toEqual(expected);
});

test("Correct non-existant empty board to reasonably full one", () => {
  const board = [];

  const boardSize = {
    width: 5,
    height: 5
  };

  const tileSet = new TileSet();
  const map = new Map(tileSet);

  const tile = map.cloneTile(1);

  const expected = [
    [tile, tile, tile, tile, tile],
    [tile, tile, tile, tile, tile],
    [tile, tile, tile, tile, tile],
    [tile, tile, tile, tile, tile],
    [tile, tile, tile, tile, tile]
  ];

  const result = map.correctBoardSizeChange(board, boardSize);
  expect(result).toEqual(expected);
});

test("Get all tiles test", () => {
  const board = [
    [
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "0",
        y: "0"
      },
      {
        id: 3,
        title: "Cacti",
        img: "cacti.png",
        background: true,
        needsDraw: false,
        frontLayer: true,
        collectable: 1,
        drawnBefore: true,
        x: "0",
        y: "1"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "0",
        y: "2"
      },
      {
        id: 3,
        title: "Cacti",
        img: "cacti.png",
        background: true,
        needsDraw: false,
        frontLayer: true,
        collectable: 1,
        drawnBefore: true,
        x: "0",
        y: "3"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "0",
        y: "4"
      },
      {
        id: 3,
        title: "Cacti",
        img: "cacti.png",
        background: true,
        needsDraw: false,
        frontLayer: true,
        collectable: 1,
        drawnBefore: true,
        x: "0",
        y: "5"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "0",
        y: "6"
      },
      {
        id: 3,
        title: "Cacti",
        img: "cacti.png",
        background: true,
        needsDraw: false,
        frontLayer: true,
        collectable: 1,
        drawnBefore: true,
        x: "0",
        y: "7"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "0",
        y: "8"
      },
      {
        id: 3,
        title: "Cacti",
        img: "cacti.png",
        background: true,
        needsDraw: false,
        frontLayer: true,
        collectable: 1,
        drawnBefore: true,
        x: "0",
        y: "9"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "0",
        y: "10"
      },
      {
        id: 3,
        title: "Cacti",
        img: "cacti.png",
        background: true,
        needsDraw: false,
        frontLayer: true,
        collectable: 1,
        drawnBefore: true,
        x: "0",
        y: "11"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "0",
        y: "12"
      },
      {
        id: 3,
        title: "Cacti",
        img: "cacti.png",
        background: true,
        needsDraw: false,
        frontLayer: true,
        collectable: 1,
        drawnBefore: true,
        x: "0",
        y: "13"
      }
    ],
    [
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "0"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "1"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "2"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "3"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "4"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "5"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "6"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "7"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "8"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "9"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "10"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "11"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "12"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "1",
        y: "13"
      }
    ],
    [
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "0"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "1"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "2"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "3"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "4"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "5"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "6"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "7"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "8"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "9"
      },
      {
        id: 2,
        title: "Fabric",
        img: "fabric.png",
        background: false,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "10"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "11"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "12"
      },
      {
        id: 1,
        title: "Sky",
        img: "sky.png",
        background: true,
        needsDraw: false,
        drawnBefore: true,
        x: "2",
        y: "13"
      }
    ]
  ];

  const tileSet = configureTileSetMock();

  const map = new Map(tileSet, null, board);

  expect(map.getAllTiles()).toMatchSnapshot();
});

test("Fix board", () => {
  const board = [[{ id: 1 }, { id: 2 }], [{ id: 2 }, { id: 1 }]];

  const tileSet = new TileSet();
  const map = new Map(tileSet);

  const tile1 = map.cloneTile(1).modify({ x: 0, y: 0 });
  const tile2 = map.cloneTile(2).modify({ x: 1, y: 0 });

  const tile3 = map.cloneTile(2).modify({ x: 0, y: 1 });
  const tile4 = map.cloneTile(1).modify({ x: 1, y: 1 });

  const expected = [[tile1, tile2], [tile3, tile4]];

  const result = map.fixBoard(board);
  expect(result).toEqual(expected);
});
