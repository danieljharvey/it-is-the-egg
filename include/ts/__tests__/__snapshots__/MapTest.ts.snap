// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get all tiles test 1`] = `
Immutable.List [
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 0,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 3,
    "title": "made up tile",
    "x": 1,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 2,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 3,
    "title": "made up tile",
    "x": 3,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 4,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 3,
    "title": "made up tile",
    "x": 5,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 6,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 3,
    "title": "made up tile",
    "x": 7,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 8,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 3,
    "title": "made up tile",
    "x": 9,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 10,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 3,
    "title": "made up tile",
    "x": 11,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 12,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 3,
    "title": "made up tile",
    "x": 13,
    "y": 0,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 0,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 1,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 2,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 3,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 4,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 5,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 6,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 7,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 8,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 9,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 10,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 11,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 12,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 13,
    "y": 1,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 0,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 1,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 2,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 3,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 4,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 5,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 6,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 7,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 8,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 9,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 2,
    "title": "made up tile",
    "x": 10,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 11,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 12,
    "y": 2,
  },
  Object {
    "action": "",
    "background": false,
    "breakable": false,
    "collectable": 0,
    "createPlayer": "",
    "dontAdd": false,
    "frontLayer": false,
    "id": 1,
    "title": "made up tile",
    "x": 13,
    "y": 2,
  },
]
`;
